pkg/metricprovider/cron_metric_provider.go

```diff
@@ -0,0 +1,455 @@
+	klog.V(6).Infof("item %v deleted", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) handleAddObj(obj interface{}) {
+	item := &workItem{obj: obj, itemType: workTypeAdded}
+	klog.V(6).Infof("item %v added", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) handleUpdateObj(oldObj, newObj interface{}) {
+	item := &workItem{obj: newObj, oldObj: oldObj, itemType: workTypeUpdated}
+	klog.V(6).Infof("item %v updated", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) reconcileEHPAWorkItem(key interface{}) error {
+	if item, ok := key.(*workItem); ok {
+		switch item.itemType {
+		case workTypeAdded:
+			p.addEHPA(item.obj)
+		case workTypeDeleted:
+			p.delEHPA(item.obj)
+		case workTypeUpdated:
+			p.updateEHPA(item.oldObj, item.obj)
+		default:
+			return fmt.Errorf("unknown workItem type, key: %+v", key)
+		}
+	}
+	return nil
+}
+
+func (p *CronMetricProvider) addEHPA(obj interface{}) {
+	ehpa, ok := obj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
+	if !ok {
+		klog.V(4).Infof("Failed to assert for obj %+v", obj)
+		return
+	}
+	scalers := GetCronScalersForEHPA(ehpa)
+
+	klog.V(6).Infof("EHPA %v starting added", klog.KObj(ehpa))
+	func() {
+		p.scalersLock.Lock()
+		defer p.scalersLock.Unlock()
+
+		for i, scaler := range scalers {
+			p.cronScalers[scaler.Name()] = scalers[i]
+		}
+	}()
+	klog.V(6).Infof("EHPA %v added finished", klog.KObj(ehpa))
+
+}
+
+func (p *CronMetricProvider) delEHPA(obj interface{}) {
+	ehpa, ok := obj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
+	if !ok {
+		klog.V(4).Infof("Failed to assert for obj %+v", obj)
+		return
+	}
+	scalers := GetCronScalersForEHPA(ehpa)
+	var names []string
+	for _, scaler := range scalers {
+		names = append(names, scaler.Name())
+	}
+
+	klog.V(6).Infof("EHPA %v starting deleted", klog.KObj(ehpa))
+	func() {
+		p.scalersLock.Lock()
+		defer p.scalersLock.Unlock()
+
+		for _, name := range names {
+			delete(p.cronScalers, name)
+		}
+	}()
+	klog.V(6).Infof("EHPA %v deleted finished", klog.KObj(ehpa))
+}
+
+func (p *CronMetricProvider) updateEHPA(oldObj, newObj interface{}) {
+	oldEhpa, ok := oldObj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
+	if !ok {
+		klog.V(4).Infof("Failed to assert for obj %+v", oldObj)
+		return
+	}
+	newEhpa, ok := newObj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
+	if !ok {
+		klog.V(4).Infof("Failed to assert for obj %+v", newEhpa)
+		return
+	}
+
+	// we update by delete old ones and add new ones
+	oldScalers := GetCronScalersForEHPA(oldEhpa)
+	newScalers := GetCronScalersForEHPA(newEhpa)
+	var oldNames []string
+	for _, oldScaler := range oldScalers {
+		oldNames = append(oldNames, oldScaler.Name())
+	}
+
+	klog.V(6).Infof("EHPA %v starting updated", klog.KObj(newEhpa))
+	func() {
+		p.scalersLock.Lock()
+		defer p.scalersLock.Unlock()
+
+		for _, name := range oldNames {
+			delete(p.cronScalers, name)
+		}
+		for _, scaler := range newScalers {
+			p.cronScalers[scaler.Name()] = scaler
```

Comment:
Do we need to save cronScalers in Provider?  
What if we just check cron status for each request and not save it in Provider.

Reply:
Yes, i removed the cache